enhancements/machine-api/aws-placement-groups.md

```diff
@@ -0,0 +1,487 @@
+
+We recommend that users leverage multiple MachineSets, each with their own Placement Group to minimise the likelihood
+of exhausting the capacity of the Placement Group.
+For example, in batch scenarios, users may wish to create a MachineSet per group of workloads. By separating each job
+into their own Placement Group, the number of instances per Placement Group may be kept small while keeping the
+benefits of grouping the workload placement in a cluster.
+
+Having spoken with AWS about the possible limitations, an engineer responded with:
+
+> The maximum number of instances in a placement group with a cluster strategy mainly depends on how many EC2 instances
+are available and close together in the same availability zone. In the past, I have been able to use more than 1,000
+c5.24xlarge instances in the same placement group with a cluster strategy.
+
+This implies that this risk may not actually be that great and that users may not hit the issues we have proposed here.
+
+#### Spread placement groups limit MachineSets to seven instances
+
+Currently, a MachineSet is limited to acting across a single Availability Zone within AWS.
+As a Spread placement group has a limit of seven instances per AZ, this then limits each MachineSet to 7 instances when
+using a Spread placement group.
+
+This may cause issues for autoscaling if users do not configure their autoscaling limits correctly.
+For example, a user may specify their MachineAutoscaler to 8 instances, even though we know it will be impossible for
+the 8th instance to launch.
+In this scenario, were the autoscaler to scale the MachineSet to 8 instances, the final instance would never launch and
+eventually the autoscaler should (dependent on configuration) scale the MachineSet back down and, if possible, attempt
+a different MachineSet.
+
+Unfortunately, there isn't currently a good way to signal to the autoscaler that this limitation exists, nor is there a
+good way to signal back to the end user that the reason for the launch failures is the seven instance limit.
+
+Using existing webhooks that inspect the Machine provider configuration, we can warn users if they create a MachineSet
+with a Spread placement group and attempt to scale it over seven instances.
+This however is not fool-proof, as the user may not have set the placement group type, or may have set it incorrectly
+and be using an existing placement group not created by MAPI.
+In this scenario the warning would not make sense, for this reason, we cannot make the warning a blocking error.
+
+We may also wish to extend the Cluster Autoscaler Operator to inspect MachineAutoscalers and their targets.
+If any MachineAutoscaler targeted a MachineSet with a spread placement group, it could warn if the maximum replica
+count is set higher than 7. This however is non-trivial as there are currently no webhooks for the Cluster Autoscaler
```

Comment:
We create this resource via the API, right?  Why can't we enforce it during creation?

Reply:
We could if we had webhooks, but we don't today and I'd rather not expand the scope of this enhancement to introduce them, it doesn't seem justified for a single validation like this.

The MachineAutoscaler is a separate CR to the MachineSet, so to check this, we need to have something lookup the MachineSet based on the object reference, then check the spec of that and report back whether it's using a spread group or not. While that's entirely possible, there's nothing to stop the end user changing the definition of their machineset during the lifecycle of the cluster, they could create the MachineAutoscaler, then configure the placement groups later